lib/experimental/sync/class-gutenberg-http-signaling-server.php

```diff
@@ -0,0 +1,343 @@
+	 */
+	private static function get_contents_from_file_descriptor( $fd ) {
+		$contents_raw = stream_get_contents( $fd );
+		$result       = array();
+		if ( $contents_raw ) {
+			$result = unserialize( $contents_raw );
+		}
+		return $result;
+	}
+
+
+	/**
+	 * Makes the file descriptor content of $fd equal to the serialization of content.
+	 * Overwrites what was previously in $fd.
+	 *
+	 * @access private
+	 * @internal
+	 *
+	 * @param resource $fd      A file descriptor.
+	 * @param array    $content Content to be serialized and written in a file descriptor.
+	 */
+	private static function save_contents_to_file_descriptor( $fd, $content ) {
+		rewind( $fd );
+		$data = serialize( $content );
+		fwrite( $fd, $data );
+		ftruncate( $fd, strlen( $data ) );
+	}
+
+	/**
+	 * Handles a wp_ajax signaling server request of client that wants to retrieve its messages.
+	 *
+	 * It returns the client a response following the
+	 * {@link https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format Event stream format}.
+	 *
+	 * ```
+	 * id: <Event ID>
+	 * retry: <Reconnection time, in ms>
+	 * event: <The type of event>
+	 * data: <The message to be sent>
+	 * ```
+	 */
+	private static function handle_message_read_request() {
+		header( 'Content-Type: text/event-stream' );
+		header( 'Cache-Control: no-cache' );
+		echo 'retry: 3000' . PHP_EOL;
+		$fd = fopen( static::$subscriber_to_messages_path, 'c+' );
+		if( ! $fd ) {
+			die( 'Could not open required file.' );
+		}
+		flock( $fd, LOCK_EX );
+		$subscriber_to_messages = static::get_contents_from_file_descriptor( $fd );
+		if ( isset( $subscriber_to_messages[ static::$subscriber_id ] ) && count( $subscriber_to_messages[ static::$subscriber_id ] ) > 0 ) {
+			echo 'id: ' . time() . PHP_EOL;
+			echo 'event: message' . PHP_EOL;
+			echo 'data: ' . wp_json_encode( $subscriber_to_messages[ static::$subscriber_id ] ) . PHP_EOL . PHP_EOL;
+			$subscriber_to_messages[ static::$subscriber_id ] = array();
+			static::save_contents_to_file_descriptor( $fd, $subscriber_to_messages );
+		} else {
+			echo PHP_EOL;
+		}
+		flock( $fd, LOCK_UN );
+		fclose( $fd );
+	}
+
+	/**
+	 * Receives a $topics_to_subscribers data-structure and an array of topics,
+	 * and returns a new $topics_to_subscribers data-structure where the current subscriber is subscribed to the topics.
+	 *
+	 * @access private
+	 * @internal
+	 *
+	 * @param array $topics_to_subscribers  Topics to subscribers data-structure.
+	 * @param array $topics                 An array of topics e.g: array( 'doc1', 'doc2' ).
+	 */
+	private static function subscribe_to_topics( $topics_to_subscribers, $topics ) {
+		foreach ( $topics as $topic ) {
+			if ( ! $topics_to_subscribers[ $topic ] ) {
+				$topics_to_subscribers[ $topic ] = array();
+			}
+			$topics_to_subscribers[ $topic ] = array_unique( array_merge( $topics_to_subscribers[ $topic ], array( static::$subscriber_id ) ) );
```

Comment:
Personally, I'd favor a more reader-focused approach:

```suggestion
			$topics_to_subscribers[ $topic ][] = static::$subscriber_id;
			$topics_to_subscribers[ $topic ]    = array_unique( $topics_to_subscribers[ $topic ] ); 
```

Of course, this is personal preference. Though I'm not sure how much it matters in this particular case, it could also be argued that it is also less server-intensive than using `array_merge`, converting to array first, etc.

Reply:
Suggestion applied.